artifactory_cleanup/rules/delete.py

```diff
@@ -92,37 +93,85 @@ class delete_empty_folder(Rule):
-            r.raise_for_status()
-
-        return []
+
+        # convert list of files to dict
+        # Source: https://stackoverflow.com/a/58917078
+
+        def nested_dict():
+            """
+            Creates a default dictionary where each value is an other default dictionary.
+            """
+            return defaultdict(nested_dict)
+
+        def default_to_regular(d):
+            """
+            Converts defaultdicts of defaultdicts to dict of dicts.
+            """
+            if isinstance(d, defaultdict):
+                d = {k: default_to_regular(v) for k, v in d.items()}
+            return d
+
+        def get_path_dict(artifacts):
+            new_path_dict = nested_dict()
+            for artifact in artifacts:
+                parts = artifact["path"].split("/")
+                if parts:
+                    marcher = new_path_dict
+                    for key in parts:
+                        # We need the repo for the root level folders. They are not in the
+                        # artifacts list
+                        marcher[key]["data"] = {"repo": artifact["repo"]}
+                        marcher = marcher[key]["children"]
+                    marcher[artifact["name"]]["data"] = artifact
+            return default_to_regular(new_path_dict)
+
+        artifact_tree = get_path_dict(result_artifact)
+
+        # Artifactory also returns the directory itself. We need to remove it from the list
+        # since that tree branch has no children assigned
+        if "." in artifact_tree:
+            del artifact_tree["."]
+
+        # Now we have a dict with all folders and files
+        # An empty folder is represented if it is a dict and does not have any keys
+
+        def get_folder_artifacts_with_no_children(item, path=""):
+
+            empty_folder_artifacts = deque()
+
+            def _add_to_del_list(key):
+                empty_folder_artifacts.append(item[key]["data"])
+                # Also delete the item from the children list to recursively delete folders
+                # upwards
+                del item[key]
+
+            for x in list(item.keys()):
+                if "type" in item[x]["data"] and item[x]["data"]["type"] == "file":
+                    continue
+                if not "path" in item[x]["data"]:
+                    # Set the path and name for root folders which were not explicitly in the
+                    # artifacts list
+                    item[x]["data"]["path"] = path
+                    item[x]["data"]["name"] = x
+                if not "children" in item[x] or len(item[x]["children"]) == 0:
+                    # This an empty folder
+                    _add_to_del_list(x)
+                else:
+                    artifacts = get_folder_artifacts_with_no_children(
+                        item[x]["children"], path=path + "/" + x if len(path) > 0 else x
+                    )
+                    if len(item[x]["children"]) == 0:
```

Comment:
There's `or len(item[x]["children"]) == 0` in `if` above, so it goes to the first `if` if `len(item[x]["children"]) == 0`, why do we need it here then?

Reply:
I mean it looks like we'll never run the block
```python
                        # just delete the whole folder since all children are empty
                        _add_to_del_list(x)
```

Reply:
It's needed here because the recursive call in line 166 may delete additional children.
And here we want to check again if all children would be deleted. Then also delete this.

Reply:
This is why I don't like mutable objects :(

Reply:
Is there a way how we can change `get_folder_artifacts_with_no_children` that it won't change their parameters?